prov/tcp/src/tcpx_cq.c

```diff
@@ -35,19 +35,66 @@
 
 #include "tcpx.h"
 
+#define TCPX_DEF_CQ_SIZE (1024)
+
 static int tcpx_cq_close(struct fid *fid)
 {
 	int ret;
-	struct util_cq *cq;
+	struct tcpx_cq *tcpx_cq;
 
-	cq = container_of(fid, struct util_cq, cq_fid.fid);
-	ret = ofi_cq_cleanup(cq);
+	tcpx_cq = container_of(fid, struct tcpx_cq, util_cq.cq_fid.fid);
+	util_buf_pool_destroy(tcpx_cq->pe_entry_pool);
+	ret = ofi_cq_cleanup(&tcpx_cq->util_cq);
 	if (ret)
 		return ret;
-	free(cq);
+
+	free(tcpx_cq);
 	return 0;
 }
 
+struct tcpx_pe_entry *tcpx_pe_entry_alloc(struct tcpx_cq *tcpx_cq)
+{
+	struct tcpx_pe_entry *pe_entry;
+
+	fastlock_acquire(&tcpx_cq->util_cq.cq_lock);
+	pe_entry = util_buf_alloc(tcpx_cq->pe_entry_pool);
+	if (!pe_entry) {
+		fastlock_release(&tcpx_cq->util_cq.cq_lock);
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,"failed to get buffer\n");
+		return NULL;
+	}
+	memset(pe_entry, 0, sizeof(*pe_entry));
+	fastlock_release(&tcpx_cq->util_cq.cq_lock);
+	return pe_entry;
+}
+
+void tcpx_pe_entry_release(struct tcpx_pe_entry *pe_entry)
+{
+	struct util_cq *cq;
+	struct tcpx_cq *tcpx_cq;
+
+	switch (pe_entry->msg_hdr.op_data) {
+	case TCPX_OP_MSG_SEND:
+		cq = pe_entry->ep->util_ep.tx_cq;
+		break;
+	case TCPX_OP_MSG_RECV:
+		cq = pe_entry->ep->util_ep.rx_cq;
+		break;
+	default:
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,"invalid pe_entry\n");
+		return;
+	}
+
+	tcpx_cq = container_of(cq, struct tcpx_cq, util_cq);
+
+	fastlock_acquire(&cq->cq_lock);
+	memset(&pe_entry->msg_hdr, 0, sizeof(pe_entry->msg_hdr));
+	dlist_remove(&pe_entry->entry);
+	memset(pe_entry, 0, sizeof(*pe_entry));
```

Comment:
We should look at avoiding blanket memset calls and only initialize whatever fields must be cleared.  If we initialize the header immediately after getting an entry, we can avoid the memset.  Likewise, do all pe_entry fields need to be 0?